backend/de.metas.business/src/main/java/org/adempiere/warehouse/interceptor/C_Order.java

```diff
@@ -0,0 +1,140 @@
+import de.metas.organization.OrgId;
+import de.metas.user.User;
+import de.metas.util.Services;
+import lombok.NonNull;
+import org.adempiere.ad.callout.annotations.Callout;
+import org.adempiere.ad.callout.annotations.CalloutMethod;
+import org.adempiere.ad.callout.spi.IProgramaticCalloutProvider;
+import org.adempiere.ad.modelvalidator.annotations.Interceptor;
+import org.adempiere.ad.modelvalidator.annotations.ModelChange;
+import org.adempiere.warehouse.WarehouseId;
+import org.adempiere.warehouse.api.IWarehouseBL;
+import org.compiere.model.I_C_BPartner;
+import org.compiere.model.I_C_BPartner_Location;
+import org.compiere.model.I_C_Order;
+import org.compiere.model.I_M_Warehouse;
+import org.compiere.model.ModelValidator;
+import org.springframework.stereotype.Component;
+
+import java.util.Optional;
+
+@Callout(I_C_Order.class)
+@Interceptor(I_C_Order.class)
+@Component
+public class C_Order
+{
+	private final IWarehouseBL warehouseBL = Services.get(IWarehouseBL.class);
+	private final IBPartnerOrgBL bPartnerOrgBL = Services.get(IBPartnerOrgBL.class);
+	private final IBPartnerDAO bPartnerDAO = Services.get(IBPartnerDAO.class);
+
+	public C_Order()
+	{
+		Services.get(IProgramaticCalloutProvider.class).registerAnnotatedCallout(this);
+	}
+
+	@CalloutMethod(columnNames = I_C_Order.COLUMNNAME_M_Warehouse_ID)
+	@ModelChange(timings = ModelValidator.TYPE_BEFORE_CHANGE, ifColumnsChanged = I_C_Order.COLUMNNAME_M_Warehouse_ID)
+	public void updateBillPartnerAndDropshipFlagFromWarehouse(@NonNull final I_C_Order order)
+	{
+		if (order.getM_Warehouse_ID() <= 0 || order.isSOTrx())
+		{
+			return; // nothing to do
+		}
+
+		final WarehouseId warehouseId = WarehouseId.ofRepoId(order.getM_Warehouse_ID());
+		final OrgId orgId = OrgId.ofRepoId(order.getAD_Org_ID());
+
+		final DocumentLocation billToAddress = getBillingDocumentForExternalBPartner(warehouseId, orgId)
+				.orElseGet(() -> getBillingDocumentForBPartnerId(BPartnerId.ofRepoId(order.getC_BPartner_ID())));
+
+		OrderDocumentLocationAdapterFactory.billLocationAdapter(order)
+				.setFrom(billToAddress);
+	}
+
+	@NonNull
+	private Optional<DocumentLocation> getBillingDocumentForExternalBPartner(@NonNull final WarehouseId warehouseId, @NonNull final OrgId orgId)
+	{
+		final I_M_Warehouse warehousesRecord = warehouseBL.getById(warehouseId);
+
+		if (!warehousesRecord.isBPartnerInvoicesWithVendors())
+		{
+			return Optional.empty();
+		}
+
+		final I_C_BPartner orgBPartner = bPartnerOrgBL.retrieveLinkedBPartner(orgId.getRepoId());
+
+		if (warehousesRecord.getC_BPartner_ID() != orgBPartner.getC_BPartner_ID())
+		{
+			return Optional.of(warehouseBL.getBPartnerBillingLocationDocument(WarehouseId.ofRepoId(warehousesRecord.getM_Warehouse_ID())));
+		}
+		return Optional.empty();
+	}
+
+	@NonNull
+	private DocumentLocation getBillingDocumentForBPartnerId(@NonNull final BPartnerId bPartnerId)
+	{
+		final I_C_BPartner bPartner = bPartnerDAO.getById(bPartnerId);
```

Comment:
i'm not sure this makes sense because absent any .`isBPartnerInvoicesWithVendors` mumbojumbo, the bill-partner is alreqady set somehow..
i think if you really want to restore the previous bill-partner, you need to find and leverage *that* other code.
but i think it's not really needed either way..